enhancements/microshift/dummy-gateway-interface-ovn.md

```diff
@@ -0,0 +1,476 @@
+see-also:
+- None
+replaces:
+- None
+superseded-by:
+- None
+---
+
+# Use Detached External Gateway Bridge of OVN-Kubernetes for MicroShift
+
+## Summary
+
+This enhancement proposes a new external gateway configuration of OVN-Kubernetes
+for MicroShift. With the new configuration, the `br-ex` will be create without
+binding any host physical interface.
+
+## Motivation
+
+MicroShift is designed to run on IoT and edge devices, it can be deployed on
+various kinds of devices with different network environment. 
+
+Today configure-ovs.sh (microshift-ovs-init.service) sets up OVN-Kubernetes
+gateway bridge `br-ex` for OVN-Kubernetes.  There has been several types of
+issues associated with it mainly due to:
+
+1. transient network interruption when it moves the node IP from uplink port to
+   gateway bridge
+
+2. certain type of network devices are not supported (e.g. wifi, tap/tun)
+
+3. it relies on (default) route info
+
+### User Stories
+
+* As an MicroShift cluster administrator, I want the installation experience of
+MicroShift is similar as a normal application. I don't want to have any
+networking interrupt during the installation.
+
+* As an MicroShift cluster administrator, I want to use a specific network
+  devices (such as wifi, tap/tun) as the node default interface.
+
+* As an MicroShift cluster administrator, I want to install MicroShift on an
+  isolated host that has no default gateway available.
+
+### Goals
+
+* Decouple the br-ex bridge from the host physical interfaces.
+* Enhance OVN-Kubernetes local gateway mode to support external gateway bridge
+  without physical NIC.
+
+### Non-Goals
+
+* Creating a new gateway mode in OVN-Kubernetes for single node cluster.
+
+## Proposal
+
+Instead of creating the br-ex bridge with the node's default interface, we
+create the br-ex without adding any host interface as its uplink port. Today, in
+order to work around the certificates issue, MicroShift adds a static kubernetes
+API server `advertise-address` (default 10.44.0.0/32) to the host `lo`
+interface. To avoid using an extra IP for the dummy gateway interface `br-ex`,
+we will use this advertiseAddress: 
+
+1. if `br-ex` doesn't exist, which means configure-ovs.sh was not invoked, the
+   CNI is not ovn-k, add advertiseAddress to `lo` interface.
+2. Otherwise, add advertiseAddress to `br-ex` interface.
+
+Currently, the local gateway mode and shared gateway mode both use the default
+gateway of the host as the default gateway of the node gateway router. But this
+is unnecessary for local gate mode. Since the pod egress traffic will leave
+OVN-Kubernetes and enter host kernel via `mp0` instead of the gateway route. So
+in local gateway mode, only the pod to hostNetwork traffic will be processed by
+the gateway router. So on the gateway router we only need to set the static
+routes to all host IPs with the ovn-k internal masquerade next-hop
+(169.254.169.4). So that the internal to localhost traffic can be forwarded to
```

Comment:
@trozet plan to fix the no default nexthop issue for both local and share gateway. IIUC, he will add all direct routes of the host to the OVN gateway router. So even if there is no default route available, the gateway route is still able to forward the traffic to localhost or the direct connect networks to br-ex. I think those routes are sufficient for local gateway mode. Since in local gw mode, all the egress traffic from GR only goes to either localhost or the other nodes of the cluster. So we don't need a default route for the gateway route.
@trozet please correct me if I'm wrong.

Reply:
I understand how that would be needed for shared gateway mode. But:

- Are we assuming the setup for local gateway mode and shared gateway mode in this regard are going to be the same even if it is not strictly needed for local gateway mode?
- I assume we might not have the default route depicted below in shared gateway mode. Do we need it here?
- Are we going to monitor the host routing table for modifications?

Reply:
Besides the secondary IP address of the node, I think we also shall also add direct routes of the host to GR. Consider we have a multiple node cluster without a default gw, and a nodeport service is backed by hostNetwork pods on other nodes.

Reply:
nodeport traffic to other nodes is load balanced at the node of destination. So for example if a pod on node A sends a packet to a nodeport on node B, the packet would leave the node with the destination IP address of node B. I don't see how that type of traffic is affected.

Reply:
Assume node A's IP is 10.89.0.3 and node-B'IP is 10.89.0.4. We have to configure the route to `10.89.0.0/24` instead of `10.89.0.3/32` in the GR of node-A, otherwise the GR don't know how to forward the traffic to node-B.

Reply:
So your example is not microshift. The example is multiple node deployment where the node's are in different subnets and there is no default gateway. In this case because we dont copy any other routes from the host we want to send the packet to the host in case it knows how to route to the other node. 

For the traffic case where we are talking about an ovnk pod -> this other node ip. I would suggest that the user turns on local gw mode if their host has a route, until we have a proper CRD so that they can create the route in OVN.

For the traffic case where we are talking about host -> service (backed by a host networked pod on another node in a different subnet)...I'm not sure even routing it back to the host would work without conntrack issues:
1. Host -> service goes into conntrack, masquerade then forwarded to OVN
2. OVN DNAT, then SNAT to host IP (in zone 0) sends it back to the host
3. Host reroutes it back out the same br-ex interface, and SNAT's to the host IP (in zone 0)

Both OVN and Host SNAT the packet to the same IP in the same zone with the same destination. I think it will break.

Reply:
Yes, putting only the host IPs in GR's route table is sufficient for Microshift. But the traffic case may happen in multi-node SNO.

In local gw mode, there is one more step between steps 2 and 3. Before being sent to the kernel, packets get to br-ex bridge. There are openflow rules to SNAT packets to `169.254.169.1`. So the original src_ip is different between steps 2 and 3. There will be different CT entries.

Additionally, in the dummy br-ex case, we have a separate IP `10.44.0.0` for br-ex, which will be used as the SNAT IP in step 2. So even the SNAT IPs are different between steps 2 and 3.

Reply:
I am bit confused on how we are considering non-microshift cases for a microshift enhancement and how we are not sticking to the requirements laid out in this enhancement.

In this specific matter, we are assuming that no change is needed because the host /32 routes are going to be there. 

If we need other types of routes, should we reflect a requirement for it?

And in any case, should we remove the default route depicted in the table if is not assumed to be needed?

Reply:
This is microshift case. But the SNO team is considering adopting this ovn-k configuration as well. Also, I don't see there is any drawback if we use the subnet instead of the host IP in the GR routing table. In my opinion, we don't really need a default route for local gw mode.

Reply:
Yeah this discussion may have broader impact than just microshift, but i think its a discussion we needed to have. Since @pliurh has tested the scenario with hairpin packets being rerouted back out br-ex and it works, I'm OK with keeping a default route back to the host.